pennylane/gradients/parameter_shift.py

```diff
@@ -327,28 +380,173 @@ def processing_fn(results):
             res = results[start : start + num_tapes] if batch_size is None else results[start]
             start = start + num_tapes
 
-            g = _evaluate_gradient(res, data, broadcast, r0, scalar_qfunc_output)
+            g = _evaluate_gradient(tape, res, data, broadcast, r0, scalar_qfunc_output)
 
             grads.append(g)
             # This clause will be hit at least once (because otherwise all gradients would have
             # been zero), providing a representative for a zero gradient to emulate its type/shape.
             zero_rep = qml.math.zeros_like(g)
 
-        for i, g in enumerate(grads):
-            # Fill in zero-valued gradients
-            if g is None:
-                grads[i] = zero_rep
-            # The following is for backwards compatibility; currently, the device stacks multiple
-            # measurement arrays, even if not the same size, resulting in a ragged array.
-            # In the future, we might want to change this so that only tuples of arrays are returned.
-            if getattr(g, "dtype", None) is np.dtype("object") and qml.math.ndim(g) > 0:
-                grads[i] = qml.math.hstack(g)
+        if qml.active_return():
+            for i, g in enumerate(grads):
+                # Fill in zero-valued gradients
+                if g is None:
+                    grads[i] = zero_rep

```

Comment:
We can merge this with the code in `_post_process_grads`

Reply:
Hmm, how? In `_post_process_grads` we have an extra `if` statement within the `for` loop.

Reply:
Something like:
```python
            for i, g in enumerate(grads):
                # Fill in zero-valued gradients
                if g is None:
                    grads[i] = zero_rep

                # The following is for backwards compatibility; currently, the device stacks multiple
                # measurement arrays, even if not the same size, resulting in a ragged array.
                # In the future, we might want to change this so that only tuples of arrays are returned.
                if (
                    not qml.active_return()
                    and getattr(g, "dtype", None) is np.dtype("object")
                    and qml.math.ndim(g) > 0
                ):
                    grads[i] = qml.math.hstack(g)
            if qml.active_return():
                res = tuple(grads)
            else:
                res = qml.math.T(qml.math.stack(grads))
```

Reply:
This would be a bad idea if `qml.active_return()` is not very performant. But I believe here it is just returning a constant, thus it should be pretty fast.

Reply:
Thanks a lot! :) Updated this part. :+1: